// @famgia/omnify-japan@2.0.3/dist/index.js
// src/locales.ts
var prefectureLabels = {
  // 北海道地方 (Hokkaido)
  hokkaido: { ja: "\u5317\u6D77\u9053", en: "Hokkaido", vi: "Hokkaido", ko: "\uD64B\uCE74\uC774\uB3C4", "zh-CN": "\u5317\u6D77\u9053", "zh-TW": "\u5317\u6D77\u9053", th: "\u0E2E\u0E2D\u0E01\u0E44\u0E01\u0E42\u0E14", es: "Hokkaido" },
  // 東北地方 (Tohoku)
  aomori: { ja: "\u9752\u68EE\u770C", en: "Aomori", vi: "Aomori", ko: "\uC544\uC624\uBAA8\uB9AC\uD604", "zh-CN": "\u9752\u68EE\u53BF", "zh-TW": "\u9752\u68EE\u7E23", th: "\u0E2D\u0E32\u0E42\u0E2D\u0E42\u0E21\u0E23\u0E34", es: "Aomori" },
  iwate: { ja: "\u5CA9\u624B\u770C", en: "Iwate", vi: "Iwate", ko: "\uC774\uC640\uD14C\uD604", "zh-CN": "\u5CA9\u624B\u53BF", "zh-TW": "\u5CA9\u624B\u7E23", th: "\u0E2D\u0E34\u0E27\u0E32\u0E40\u0E15\u0E30", es: "Iwate" },
  miyagi: { ja: "\u5BAE\u57CE\u770C", en: "Miyagi", vi: "Miyagi", ko: "\uBBF8\uC57C\uAE30\uD604", "zh-CN": "\u5BAB\u57CE\u53BF", "zh-TW": "\u5BAE\u57CE\u7E23", th: "\u0E21\u0E34\u0E22\u0E32\u0E07\u0E34", es: "Miyagi" },
  akita: { ja: "\u79CB\u7530\u770C", en: "Akita", vi: "Akita", ko: "\uC544\uD0A4\uD0C0\uD604", "zh-CN": "\u79CB\u7530\u53BF", "zh-TW": "\u79CB\u7530\u7E23", th: "\u0E2D\u0E32\u0E01\u0E34\u0E15\u0E30", es: "Akita" },
  yamagata: { ja: "\u5C71\u5F62\u770C", en: "Yamagata", vi: "Yamagata", ko: "\uC57C\uB9C8\uAC00\uD0C0\uD604", "zh-CN": "\u5C71\u5F62\u53BF", "zh-TW": "\u5C71\u5F62\u7E23", th: "\u0E22\u0E32\u0E21\u0E32\u0E01\u0E32\u0E15\u0E30", es: "Yamagata" },
  fukushima: { ja: "\u798F\u5CF6\u770C", en: "Fukushima", vi: "Fukushima", ko: "\uD6C4\uCFE0\uC2DC\uB9C8\uD604", "zh-CN": "\u798F\u5C9B\u53BF", "zh-TW": "\u798F\u5CF6\u7E23", th: "\u0E1F\u0E38\u0E01\u0E38\u0E0A\u0E34\u0E21\u0E30", es: "Fukushima" },
  // 関東地方 (Kanto)
  ibaraki: { ja: "\u8328\u57CE\u770C", en: "Ibaraki", vi: "Ibaraki", ko: "\uC774\uBC14\uB77C\uD0A4\uD604", "zh-CN": "\u8328\u57CE\u53BF", "zh-TW": "\u8328\u57CE\u7E23", th: "\u0E2D\u0E34\u0E1A\u0E32\u0E23\u0E32\u0E01\u0E34", es: "Ibaraki" },
  tochigi: { ja: "\u6803\u6728\u770C", en: "Tochigi", vi: "Tochigi", ko: "\uB3C4\uCE58\uAE30\uD604", "zh-CN": "\u6803\u6728\u53BF", "zh-TW": "\u6803\u6728\u7E23", th: "\u0E42\u0E17\u0E08\u0E34\u0E07\u0E34", es: "Tochigi" },
  gunma: { ja: "\u7FA4\u99AC\u770C", en: "Gunma", vi: "Gunma", ko: "\uAD70\uB9C8\uD604", "zh-CN": "\u7FA4\u9A6C\u53BF", "zh-TW": "\u7FA4\u99AC\u7E23", th: "\u0E01\u0E38\u0E19\u0E21\u0E30", es: "Gunma" },
  saitama: { ja: "\u57FC\u7389\u770C", en: "Saitama", vi: "Saitama", ko: "\uC0AC\uC774\uD0C0\uB9C8\uD604", "zh-CN": "\u57FC\u7389\u53BF", "zh-TW": "\u57FC\u7389\u7E23", th: "\u0E44\u0E0B\u0E15\u0E32\u0E21\u0E30", es: "Saitama" },
  chiba: { ja: "\u5343\u8449\u770C", en: "Chiba", vi: "Chiba", ko: "\uCE58\uBC14\uD604", "zh-CN": "\u5343\u53F6\u53BF", "zh-TW": "\u5343\u8449\u7E23", th: "\u0E08\u0E34\u0E1A\u0E30", es: "Chiba" },
  tokyo: { ja: "\u6771\u4EAC\u90FD", en: "Tokyo", vi: "Tokyo", ko: "\uB3C4\uCFC4\uB3C4", "zh-CN": "\u4E1C\u4EAC\u90FD", "zh-TW": "\u6771\u4EAC\u90FD", th: "\u0E42\u0E15\u0E40\u0E01\u0E35\u0E22\u0E27", es: "Tokio" },
  kanagawa: { ja: "\u795E\u5948\u5DDD\u770C", en: "Kanagawa", vi: "Kanagawa", ko: "\uAC00\uB098\uAC00\uC640\uD604", "zh-CN": "\u795E\u5948\u5DDD\u53BF", "zh-TW": "\u795E\u5948\u5DDD\u7E23", th: "\u0E04\u0E32\u0E19\u0E32\u0E01\u0E32\u0E27\u0E30", es: "Kanagawa" },
  // 中部地方 (Chubu)
  niigata: { ja: "\u65B0\u6F5F\u770C", en: "Niigata", vi: "Niigata", ko: "\uB2C8\uAC00\uD0C0\uD604", "zh-CN": "\u65B0\u6F5F\u53BF", "zh-TW": "\u65B0\u6F5F\u7E23", th: "\u0E19\u0E35\u0E07\u0E32\u0E15\u0E30", es: "Niigata" },
  toyama: { ja: "\u5BCC\u5C71\u770C", en: "Toyama", vi: "Toyama", ko: "\uB3C4\uC57C\uB9C8\uD604", "zh-CN": "\u5BCC\u5C71\u53BF", "zh-TW": "\u5BCC\u5C71\u7E23", th: "\u0E42\u0E17\u0E22\u0E32\u0E21\u0E30", es: "Toyama" },
  ishikawa: { ja: "\u77F3\u5DDD\u770C", en: "Ishikawa", vi: "Ishikawa", ko: "\uC774\uC2DC\uCE74\uC640\uD604", "zh-CN": "\u77F3\u5DDD\u53BF", "zh-TW": "\u77F3\u5DDD\u7E23", th: "\u0E2D\u0E34\u0E0A\u0E34\u0E01\u0E32\u0E27\u0E30", es: "Ishikawa" },
  fukui: { ja: "\u798F\u4E95\u770C", en: "Fukui", vi: "Fukui", ko: "\uD6C4\uCFE0\uC774\uD604", "zh-CN": "\u798F\u4E95\u53BF", "zh-TW": "\u798F\u4E95\u7E23", th: "\u0E1F\u0E38\u0E01\u0E38\u0E2D\u0E34", es: "Fukui" },
  yamanashi: { ja: "\u5C71\u68A8\u770C", en: "Yamanashi", vi: "Yamanashi", ko: "\uC57C\uB9C8\uB098\uC2DC\uD604", "zh-CN": "\u5C71\u68A8\u53BF", "zh-TW": "\u5C71\u68A8\u7E23", th: "\u0E22\u0E32\u0E21\u0E32\u0E19\u0E32\u0E0A\u0E34", es: "Yamanashi" },
  nagano: { ja: "\u9577\u91CE\u770C", en: "Nagano", vi: "Nagano", ko: "\uB098\uAC00\uB178\uD604", "zh-CN": "\u957F\u91CE\u53BF", "zh-TW": "\u9577\u91CE\u7E23", th: "\u0E19\u0E32\u0E01\u0E32\u0E42\u0E19\u0E30", es: "Nagano" },
  gifu: { ja: "\u5C90\u961C\u770C", en: "Gifu", vi: "Gifu", ko: "\uAE30\uD6C4\uD604", "zh-CN": "\u5C90\u961C\u53BF", "zh-TW": "\u5C90\u961C\u7E23", th: "\u0E01\u0E34\u0E1F\u0E38", es: "Gifu" },
  shizuoka: { ja: "\u9759\u5CA1\u770C", en: "Shizuoka", vi: "Shizuoka", ko: "\uC2DC\uC988\uC624\uCE74\uD604", "zh-CN": "\u9759\u5188\u53BF", "zh-TW": "\u975C\u5CA1\u7E23", th: "\u0E0A\u0E34\u0E0B\u0E39\u0E42\u0E2D\u0E01\u0E30", es: "Shizuoka" },
  aichi: { ja: "\u611B\u77E5\u770C", en: "Aichi", vi: "Aichi", ko: "\uC544\uC774\uCE58\uD604", "zh-CN": "\u7231\u77E5\u53BF", "zh-TW": "\u611B\u77E5\u7E23", th: "\u0E44\u0E2D\u0E08\u0E34", es: "Aichi" },
  // 近畿地方 (Kinki/Kansai)
  mie: { ja: "\u4E09\u91CD\u770C", en: "Mie", vi: "Mie", ko: "\uBBF8\uC5D0\uD604", "zh-CN": "\u4E09\u91CD\u53BF", "zh-TW": "\u4E09\u91CD\u7E23", th: "\u0E21\u0E34\u0E40\u0E2D\u0E30", es: "Mie" },
  shiga: { ja: "\u6ECB\u8CC0\u770C", en: "Shiga", vi: "Shiga", ko: "\uC2DC\uAC00\uD604", "zh-CN": "\u6ECB\u8D3A\u53BF", "zh-TW": "\u6ECB\u8CC0\u7E23", th: "\u0E0A\u0E34\u0E07\u0E30", es: "Shiga" },
  kyoto: { ja: "\u4EAC\u90FD\u5E9C", en: "Kyoto", vi: "Kyoto", ko: "\uAD50\uD1A0\uBD80", "zh-CN": "\u4EAC\u90FD\u5E9C", "zh-TW": "\u4EAC\u90FD\u5E9C", th: "\u0E40\u0E01\u0E35\u0E22\u0E27\u0E42\u0E15", es: "Kioto" },
  osaka: { ja: "\u5927\u962A\u5E9C", en: "Osaka", vi: "Osaka", ko: "\uC624\uC0AC\uCE74\uBD80", "zh-CN": "\u5927\u962A\u5E9C", "zh-TW": "\u5927\u962A\u5E9C", th: "\u0E42\u0E2D\u0E0B\u0E32\u0E01\u0E32", es: "Osaka" },
  hyogo: { ja: "\u5175\u5EAB\u770C", en: "Hyogo", vi: "Hyogo", ko: "\uD6A8\uACE0\uD604", "zh-CN": "\u5175\u5E93\u53BF", "zh-TW": "\u5175\u5EAB\u7E23", th: "\u0E40\u0E2E\u0E35\u0E22\u0E27\u0E42\u0E07\u0E30", es: "Hyogo" },
  nara: { ja: "\u5948\u826F\u770C", en: "Nara", vi: "Nara", ko: "\uB098\uB77C\uD604", "zh-CN": "\u5948\u826F\u53BF", "zh-TW": "\u5948\u826F\u7E23", th: "\u0E19\u0E32\u0E23\u0E30", es: "Nara" },
  wakayama: { ja: "\u548C\u6B4C\u5C71\u770C", en: "Wakayama", vi: "Wakayama", ko: "\uC640\uCE74\uC57C\uB9C8\uD604", "zh-CN": "\u548C\u6B4C\u5C71\u53BF", "zh-TW": "\u548C\u6B4C\u5C71\u7E23", th: "\u0E27\u0E32\u0E01\u0E32\u0E22\u0E32\u0E21\u0E30", es: "Wakayama" },
  // 中国地方 (Chugoku)
  tottori: { ja: "\u9CE5\u53D6\u770C", en: "Tottori", vi: "Tottori", ko: "\uB3D7\uD1A0\uB9AC\uD604", "zh-CN": "\u9E1F\u53D6\u53BF", "zh-TW": "\u9CE5\u53D6\u7E23", th: "\u0E17\u0E47\u0E2D\u0E15\u0E42\u0E15\u0E23\u0E34", es: "Tottori" },
  shimane: { ja: "\u5CF6\u6839\u770C", en: "Shimane", vi: "Shimane", ko: "\uC2DC\uB9C8\uB124\uD604", "zh-CN": "\u5C9B\u6839\u53BF", "zh-TW": "\u5CF6\u6839\u7E23", th: "\u0E0A\u0E34\u0E21\u0E32\u0E40\u0E19\u0E30", es: "Shimane" },
  okayama: { ja: "\u5CA1\u5C71\u770C", en: "Okayama", vi: "Okayama", ko: "\uC624\uCE74\uC57C\uB9C8\uD604", "zh-CN": "\u5188\u5C71\u53BF", "zh-TW": "\u5CA1\u5C71\u7E23", th: "\u0E42\u0E2D\u0E04\u0E32\u0E22\u0E32\u0E21\u0E30", es: "Okayama" },
  hiroshima: { ja: "\u5E83\u5CF6\u770C", en: "Hiroshima", vi: "Hiroshima", ko: "\uD788\uB85C\uC2DC\uB9C8\uD604", "zh-CN": "\u5E7F\u5C9B\u53BF", "zh-TW": "\u5EE3\u5CF6\u7E23", th: "\u0E2E\u0E34\u0E42\u0E23\u0E0A\u0E34\u0E21\u0E30", es: "Hiroshima" },
  yamaguchi: { ja: "\u5C71\u53E3\u770C", en: "Yamaguchi", vi: "Yamaguchi", ko: "\uC57C\uB9C8\uAD6C\uCE58\uD604", "zh-CN": "\u5C71\u53E3\u53BF", "zh-TW": "\u5C71\u53E3\u7E23", th: "\u0E22\u0E32\u0E21\u0E32\u0E01\u0E39\u0E08\u0E34", es: "Yamaguchi" },
  // 四国地方 (Shikoku)
  tokushima: { ja: "\u5FB3\u5CF6\u770C", en: "Tokushima", vi: "Tokushima", ko: "\uB3C4\uCFE0\uC2DC\uB9C8\uD604", "zh-CN": "\u5FB7\u5C9B\u53BF", "zh-TW": "\u5FB7\u5CF6\u7E23", th: "\u0E42\u0E17\u0E01\u0E38\u0E0A\u0E34\u0E21\u0E30", es: "Tokushima" },
  kagawa: { ja: "\u9999\u5DDD\u770C", en: "Kagawa", vi: "Kagawa", ko: "\uAC00\uAC00\uC640\uD604", "zh-CN": "\u9999\u5DDD\u53BF", "zh-TW": "\u9999\u5DDD\u7E23", th: "\u0E04\u0E32\u0E01\u0E32\u0E27\u0E30", es: "Kagawa" },
  ehime: { ja: "\u611B\u5A9B\u770C", en: "Ehime", vi: "Ehime", ko: "\uC5D0\uD788\uBA54\uD604", "zh-CN": "\u7231\u5A9B\u53BF", "zh-TW": "\u611B\u5A9B\u7E23", th: "\u0E40\u0E2D\u0E2E\u0E34\u0E40\u0E21\u0E30", es: "Ehime" },
  kochi: { ja: "\u9AD8\u77E5\u770C", en: "Kochi", vi: "Kochi", ko: "\uACE0\uCE58\uD604", "zh-CN": "\u9AD8\u77E5\u53BF", "zh-TW": "\u9AD8\u77E5\u7E23", th: "\u0E42\u0E04\u0E08\u0E34", es: "Kochi" },
  // 九州・沖縄地方 (Kyushu & Okinawa)
  fukuoka: { ja: "\u798F\u5CA1\u770C", en: "Fukuoka", vi: "Fukuoka", ko: "\uD6C4\uCFE0\uC624\uCE74\uD604", "zh-CN": "\u798F\u5188\u53BF", "zh-TW": "\u798F\u5CA1\u7E23", th: "\u0E1F\u0E38\u0E01\u0E38\u0E42\u0E2D\u0E01\u0E30", es: "Fukuoka" },
  saga: { ja: "\u4F50\u8CC0\u770C", en: "Saga", vi: "Saga", ko: "\uC0AC\uAC00\uD604", "zh-CN": "\u4F50\u8D3A\u53BF", "zh-TW": "\u4F50\u8CC0\u7E23", th: "\u0E0B\u0E32\u0E07\u0E30", es: "Saga" },
  nagasaki: { ja: "\u9577\u5D0E\u770C", en: "Nagasaki", vi: "Nagasaki", ko: "\uB098\uAC00\uC0AC\uD0A4\uD604", "zh-CN": "\u957F\u5D0E\u53BF", "zh-TW": "\u9577\u5D0E\u7E23", th: "\u0E19\u0E32\u0E07\u0E32\u0E0B\u0E32\u0E01\u0E34", es: "Nagasaki" },
  kumamoto: { ja: "\u718A\u672C\u770C", en: "Kumamoto", vi: "Kumamoto", ko: "\uAD6C\uB9C8\uBAA8\uD1A0\uD604", "zh-CN": "\u718A\u672C\u53BF", "zh-TW": "\u718A\u672C\u7E23", th: "\u0E04\u0E38\u0E21\u0E32\u0E42\u0E21\u0E42\u0E15\u0E30", es: "Kumamoto" },
  oita: { ja: "\u5927\u5206\u770C", en: "Oita", vi: "Oita", ko: "\uC624\uC774\uD0C0\uD604", "zh-CN": "\u5927\u5206\u53BF", "zh-TW": "\u5927\u5206\u7E23", th: "\u0E42\u0E2D\u0E2D\u0E34\u0E15\u0E30", es: "Oita" },
  miyazaki: { ja: "\u5BAE\u5D0E\u770C", en: "Miyazaki", vi: "Miyazaki", ko: "\uBBF8\uC57C\uC790\uD0A4\uD604", "zh-CN": "\u5BAB\u5D0E\u53BF", "zh-TW": "\u5BAE\u5D0E\u7E23", th: "\u0E21\u0E34\u0E22\u0E32\u0E0B\u0E32\u0E01\u0E34", es: "Miyazaki" },
  kagoshima: { ja: "\u9E7F\u5150\u5CF6\u770C", en: "Kagoshima", vi: "Kagoshima", ko: "\uAC00\uACE0\uC2DC\uB9C8\uD604", "zh-CN": "\u9E7F\u513F\u5C9B\u53BF", "zh-TW": "\u9E7F\u5152\u5CF6\u7E23", th: "\u0E04\u0E32\u0E42\u0E01\u0E0A\u0E34\u0E21\u0E30", es: "Kagoshima" },
  okinawa: { ja: "\u6C96\u7E04\u770C", en: "Okinawa", vi: "Okinawa", ko: "\uC624\uD0A4\uB098\uC640\uD604", "zh-CN": "\u51B2\u7EF3\u53BF", "zh-TW": "\u6C96\u7E69\u7E23", th: "\u0E42\u0E2D\u0E01\u0E34\u0E19\u0E32\u0E27\u0E30", es: "Okinawa" }
};
var prefectureAreaLabels = {
  hokkaido: { ja: "\u5317\u6D77\u9053\u5730\u65B9", en: "Hokkaido", vi: "V\xF9ng Hokkaido", ko: "\uD64B\uCE74\uC774\uB3C4 \uC9C0\uBC29", "zh-CN": "\u5317\u6D77\u9053\u5730\u533A", "zh-TW": "\u5317\u6D77\u9053\u5730\u5340", th: "\u0E20\u0E32\u0E04\u0E2E\u0E2D\u0E01\u0E44\u0E01\u0E42\u0E14", es: "Hokkaido" },
  tohoku: { ja: "\u6771\u5317\u5730\u65B9", en: "Tohoku", vi: "V\xF9ng Tohoku", ko: "\uB3C4\uD638\uCFE0 \uC9C0\uBC29", "zh-CN": "\u4E1C\u5317\u5730\u533A", "zh-TW": "\u6771\u5317\u5730\u5340", th: "\u0E20\u0E32\u0E04\u0E42\u0E17\u0E42\u0E2E\u0E01\u0E38", es: "Tohoku" },
  kanto: { ja: "\u95A2\u6771\u5730\u65B9", en: "Kanto", vi: "V\xF9ng Kanto", ko: "\uAC04\uD1A0 \uC9C0\uBC29", "zh-CN": "\u5173\u4E1C\u5730\u533A", "zh-TW": "\u95DC\u6771\u5730\u5340", th: "\u0E20\u0E32\u0E04\u0E04\u0E31\u0E19\u0E42\u0E15", es: "Kanto" },
  chubu: { ja: "\u4E2D\u90E8\u5730\u65B9", en: "Chubu", vi: "V\xF9ng Chubu", ko: "\uC8FC\uBD80 \uC9C0\uBC29", "zh-CN": "\u4E2D\u90E8\u5730\u533A", "zh-TW": "\u4E2D\u90E8\u5730\u5340", th: "\u0E20\u0E32\u0E04\u0E0A\u0E39\u0E1A\u0E38", es: "Chubu" },
  kinki: { ja: "\u8FD1\u757F\u5730\u65B9", en: "Kinki/Kansai", vi: "V\xF9ng Kinki", ko: "\uAE34\uD0A4 \uC9C0\uBC29", "zh-CN": "\u8FD1\u757F\u5730\u533A", "zh-TW": "\u8FD1\u757F\u5730\u5340", th: "\u0E20\u0E32\u0E04\u0E04\u0E34\u0E19\u0E01\u0E34", es: "Kinki/Kansai" },
  chugoku: { ja: "\u4E2D\u56FD\u5730\u65B9", en: "Chugoku", vi: "V\xF9ng Chugoku", ko: "\uC8FC\uACE0\uCFE0 \uC9C0\uBC29", "zh-CN": "\u4E2D\u56FD\u5730\u533A", "zh-TW": "\u4E2D\u570B\u5730\u5340", th: "\u0E20\u0E32\u0E04\u0E0A\u0E39\u0E42\u0E07\u0E01\u0E38", es: "Chugoku" },
  shikoku: { ja: "\u56DB\u56FD\u5730\u65B9", en: "Shikoku", vi: "V\xF9ng Shikoku", ko: "\uC2DC\uCF54\uCFE0 \uC9C0\uBC29", "zh-CN": "\u56DB\u56FD\u5730\u533A", "zh-TW": "\u56DB\u570B\u5730\u5340", th: "\u0E20\u0E32\u0E04\u0E0A\u0E34\u0E42\u0E01\u0E01\u0E38", es: "Shikoku" },
  kyushu: { ja: "\u4E5D\u5DDE\u30FB\u6C96\u7E04\u5730\u65B9", en: "Kyushu & Okinawa", vi: "V\xF9ng Kyushu & Okinawa", ko: "\uADDC\uC288\u30FB\uC624\uD0A4\uB098\uC640 \uC9C0\uBC29", "zh-CN": "\u4E5D\u5DDE\u30FB\u51B2\u7EF3\u5730\u533A", "zh-TW": "\u4E5D\u5DDE\u30FB\u6C96\u7E69\u5730\u5340", th: "\u0E20\u0E32\u0E04\u0E04\u0E34\u0E27\u0E0A\u0E39\u0E41\u0E25\u0E30\u0E42\u0E2D\u0E01\u0E34\u0E19\u0E32\u0E27\u0E30", es: "Kyushu y Okinawa" }
};
var bankAccountTypeLabels = {
  "1": { ja: "\u666E\u901A", en: "Ordinary", vi: "Th\u01B0\u1EDDng", ko: "\uBCF4\uD1B5", "zh-CN": "\u666E\u901A", "zh-TW": "\u666E\u901A", th: "\u0E1B\u0E01\u0E15\u0E34", es: "Ordinario" },
  "2": { ja: "\u5F53\u5EA7", en: "Checking", vi: "Thanh to\xE1n", ko: "\uB2F9\uC88C", "zh-CN": "\u6D3B\u671F", "zh-TW": "\u6D3B\u671F", th: "\u0E01\u0E23\u0E30\u0E41\u0E2A\u0E23\u0E32\u0E22\u0E27\u0E31\u0E19", es: "Corriente" },
  "4": { ja: "\u8CAF\u84C4", en: "Savings", vi: "Ti\u1EBFt ki\u1EC7m", ko: "\uC800\uCD95", "zh-CN": "\u50A8\u84C4", "zh-TW": "\u5132\u84C4", th: "\u0E2D\u0E2D\u0E21\u0E17\u0E23\u0E31\u0E1E\u0E22\u0E4C", es: "Ahorro" }
};
var enumDisplayNames = {
  Prefecture: { ja: "\u90FD\u9053\u5E9C\u770C", en: "Prefecture", vi: "T\u1EC9nh/Th\xE0nh ph\u1ED1", ko: "\uB3C4\uB3C4\uBD80\uD604", "zh-CN": "\u90FD\u9053\u5E9C\u53BF", "zh-TW": "\u90FD\u9053\u5E9C\u7E23", th: "\u0E08\u0E31\u0E07\u0E2B\u0E27\u0E31\u0E14", es: "Prefectura" },
  PrefectureCode: { ja: "\u90FD\u9053\u5E9C\u770C\u30B3\u30FC\u30C9", en: "Prefecture Code", vi: "M\xE3 t\u1EC9nh/th\xE0nh", ko: "\uB3C4\uB3C4\uBD80\uD604 \uCF54\uB4DC", "zh-CN": "\u90FD\u9053\u5E9C\u53BF\u4EE3\u7801", "zh-TW": "\u90FD\u9053\u5E9C\u7E23\u4EE3\u78BC", th: "\u0E23\u0E2B\u0E31\u0E2A\u0E08\u0E31\u0E07\u0E2B\u0E27\u0E31\u0E14", es: "C\xF3digo de Prefectura" },
  BankAccountType: { ja: "\u53E3\u5EA7\u7A2E\u5225", en: "Account Type", vi: "Lo\u1EA1i t\xE0i kho\u1EA3n", ko: "\uACC4\uC88C \uC885\uB958", "zh-CN": "\u8D26\u6237\u7C7B\u578B", "zh-TW": "\u5E33\u6236\u985E\u578B", th: "\u0E1B\u0E23\u0E30\u0E40\u0E20\u0E17\u0E1A\u0E31\u0E0D\u0E0A\u0E35", es: "Tipo de Cuenta" }
};
var fieldLabels = {
  // Address fields
  postalCode: { ja: "\u90F5\u4FBF\u756A\u53F7", en: "Postal Code", vi: "M\xE3 b\u01B0u \u0111i\u1EC7n", ko: "\uC6B0\uD3B8\uBC88\uD638", "zh-CN": "\u90AE\u653F\u7F16\u7801", "zh-TW": "\u90F5\u905E\u5340\u865F", th: "\u0E23\u0E2B\u0E31\u0E2A\u0E44\u0E1B\u0E23\u0E29\u0E13\u0E35\u0E22\u0E4C", es: "C\xF3digo Postal" },
  prefecture: { ja: "\u90FD\u9053\u5E9C\u770C", en: "Prefecture", vi: "T\u1EC9nh/Th\xE0nh ph\u1ED1", ko: "\uB3C4\uB3C4\uBD80\uD604", "zh-CN": "\u90FD\u9053\u5E9C\u53BF", "zh-TW": "\u90FD\u9053\u5E9C\u7E23", th: "\u0E08\u0E31\u0E07\u0E2B\u0E27\u0E31\u0E14", es: "Prefectura" },
  address1: { ja: "\u5E02\u533A\u753A\u6751", en: "City/Ward", vi: "Th\xE0nh ph\u1ED1/Qu\u1EADn", ko: "\uC2DC\uAD6C\uC815\uCD0C", "zh-CN": "\u5E02\u533A\u753A\u6751", "zh-TW": "\u5E02\u5340\u753A\u6751", th: "\u0E40\u0E21\u0E37\u0E2D\u0E07/\u0E40\u0E02\u0E15", es: "Ciudad/Distrito" },
  address2: { ja: "\u753A\u540D\u30FB\u756A\u5730", en: "Street Address", vi: "\u0110\u1ECBa ch\u1EC9", ko: "\uB3D9\u30FB\uBC88\uC9C0", "zh-CN": "\u8857\u9053\u5730\u5740", "zh-TW": "\u8857\u9053\u5730\u5740", th: "\u0E17\u0E35\u0E48\u0E2D\u0E22\u0E39\u0E48", es: "Direcci\xF3n" },
  address3: { ja: "\u5EFA\u7269\u540D\u30FB\u90E8\u5C4B\u756A\u53F7", en: "Building/Room", vi: "T\xF2a nh\xE0/Ph\xF2ng", ko: "\uAC74\uBB3C\uBA85\u30FB\uD638\uC2E4", "zh-CN": "\u5EFA\u7B51\u7269/\u623F\u95F4", "zh-TW": "\u5EFA\u7BC9\u7269/\u623F\u9593", th: "\u0E2D\u0E32\u0E04\u0E32\u0E23/\u0E2B\u0E49\u0E2D\u0E07", es: "Edificio/Habitaci\xF3n" },
  // Name fields
  lastname: { ja: "\u59D3", en: "Last Name", vi: "H\u1ECD", ko: "\uC131", "zh-CN": "\u59D3", "zh-TW": "\u59D3", th: "\u0E19\u0E32\u0E21\u0E2A\u0E01\u0E38\u0E25", es: "Apellido" },
  firstname: { ja: "\u540D", en: "First Name", vi: "T\xEAn", ko: "\uC774\uB984", "zh-CN": "\u540D", "zh-TW": "\u540D", th: "\u0E0A\u0E37\u0E48\u0E2D", es: "Nombre" },
  kanaLastname: { ja: "\u59D3\uFF08\u30AB\u30CA\uFF09", en: "Last Name (Kana)", vi: "H\u1ECD (Kana)", ko: "\uC131 (\uAC00\uB098)", "zh-CN": "\u59D3\uFF08\u5047\u540D\uFF09", "zh-TW": "\u59D3\uFF08\u5047\u540D\uFF09", th: "\u0E19\u0E32\u0E21\u0E2A\u0E01\u0E38\u0E25 (\u0E04\u0E32\u0E19\u0E30)", es: "Apellido (Kana)" },
  kanaFirstname: { ja: "\u540D\uFF08\u30AB\u30CA\uFF09", en: "First Name (Kana)", vi: "T\xEAn (Kana)", ko: "\uC774\uB984 (\uAC00\uB098)", "zh-CN": "\u540D\uFF08\u5047\u540D\uFF09", "zh-TW": "\u540D\uFF08\u5047\u540D\uFF09", th: "\u0E0A\u0E37\u0E48\u0E2D (\u0E04\u0E32\u0E19\u0E30)", es: "Nombre (Kana)" },
  // Phone
  phone: { ja: "\u96FB\u8A71\u756A\u53F7", en: "Phone Number", vi: "S\u1ED1 \u0111i\u1EC7n tho\u1EA1i", ko: "\uC804\uD654\uBC88\uD638", "zh-CN": "\u7535\u8BDD\u53F7\u7801", "zh-TW": "\u96FB\u8A71\u865F\u78BC", th: "\u0E2B\u0E21\u0E32\u0E22\u0E40\u0E25\u0E02\u0E42\u0E17\u0E23\u0E28\u0E31\u0E1E\u0E17\u0E4C", es: "N\xFAmero de Tel\xE9fono" },
  // Bank account fields
  bankCode: { ja: "\u9280\u884C\u30B3\u30FC\u30C9", en: "Bank Code", vi: "M\xE3 ng\xE2n h\xE0ng", ko: "\uC740\uD589 \uCF54\uB4DC", "zh-CN": "\u94F6\u884C\u4EE3\u7801", "zh-TW": "\u9280\u884C\u4EE3\u78BC", th: "\u0E23\u0E2B\u0E31\u0E2A\u0E18\u0E19\u0E32\u0E04\u0E32\u0E23", es: "C\xF3digo de Banco" },
  bankName: { ja: "\u9280\u884C\u540D", en: "Bank Name", vi: "T\xEAn ng\xE2n h\xE0ng", ko: "\uC740\uD589\uBA85", "zh-CN": "\u94F6\u884C\u540D\u79F0", "zh-TW": "\u9280\u884C\u540D\u7A31", th: "\u0E0A\u0E37\u0E48\u0E2D\u0E18\u0E19\u0E32\u0E04\u0E32\u0E23", es: "Nombre del Banco" },
  branchCode: { ja: "\u652F\u5E97\u30B3\u30FC\u30C9", en: "Branch Code", vi: "M\xE3 chi nh\xE1nh", ko: "\uC9C0\uC810 \uCF54\uB4DC", "zh-CN": "\u652F\u5E97\u4EE3\u7801", "zh-TW": "\u652F\u5E97\u4EE3\u78BC", th: "\u0E23\u0E2B\u0E31\u0E2A\u0E2A\u0E32\u0E02\u0E32", es: "C\xF3digo de Sucursal" },
  branchName: { ja: "\u652F\u5E97\u540D", en: "Branch Name", vi: "T\xEAn chi nh\xE1nh", ko: "\uC9C0\uC810\uBA85", "zh-CN": "\u652F\u5E97\u540D\u79F0", "zh-TW": "\u652F\u5E97\u540D\u7A31", th: "\u0E0A\u0E37\u0E48\u0E2D\u0E2A\u0E32\u0E02\u0E32", es: "Nombre de Sucursal" },
  accountType: { ja: "\u53E3\u5EA7\u7A2E\u5225", en: "Account Type", vi: "Lo\u1EA1i t\xE0i kho\u1EA3n", ko: "\uACC4\uC88C \uC885\uB958", "zh-CN": "\u8D26\u6237\u7C7B\u578B", "zh-TW": "\u5E33\u6236\u985E\u578B", th: "\u0E1B\u0E23\u0E30\u0E40\u0E20\u0E17\u0E1A\u0E31\u0E0D\u0E0A\u0E35", es: "Tipo de Cuenta" },
  accountNumber: { ja: "\u53E3\u5EA7\u756A\u53F7", en: "Account Number", vi: "S\u1ED1 t\xE0i kho\u1EA3n", ko: "\uACC4\uC88C\uBC88\uD638", "zh-CN": "\u8D26\u53F7", "zh-TW": "\u5E33\u865F", th: "\u0E2B\u0E21\u0E32\u0E22\u0E40\u0E25\u0E02\u0E1A\u0E31\u0E0D\u0E0A\u0E35", es: "N\xFAmero de Cuenta" },
  accountHolder: { ja: "\u53E3\u5EA7\u540D\u7FA9", en: "Account Holder", vi: "Ch\u1EE7 t\xE0i kho\u1EA3n", ko: "\uC608\uAE08\uC8FC", "zh-CN": "\u8D26\u6237\u540D", "zh-TW": "\u5E33\u6236\u540D", th: "\u0E0A\u0E37\u0E48\u0E2D\u0E1A\u0E31\u0E0D\u0E0A\u0E35", es: "Titular de la Cuenta" }
};
var placeholders = {
  postalCode: { ja: "\u4F8B\uFF1A123-4567", en: "e.g. 123-4567", vi: "VD: 123-4567", ko: "\uC608: 123-4567", "zh-CN": "\u4F8B\uFF1A123-4567", "zh-TW": "\u4F8B\uFF1A123-4567", th: "\u0E40\u0E0A\u0E48\u0E19 123-4567", es: "ej. 123-4567" },
  prefecture: { ja: "\u90FD\u9053\u5E9C\u770C\u3092\u9078\u629E", en: "Select prefecture", vi: "Ch\u1ECDn t\u1EC9nh/th\xE0nh", ko: "\uB3C4\uB3C4\uBD80\uD604 \uC120\uD0DD", "zh-CN": "\u9009\u62E9\u90FD\u9053\u5E9C\u53BF", "zh-TW": "\u9078\u64C7\u90FD\u9053\u5E9C\u7E23", th: "\u0E40\u0E25\u0E37\u0E2D\u0E01\u0E08\u0E31\u0E07\u0E2B\u0E27\u0E31\u0E14", es: "Seleccionar prefectura" },
  address1: { ja: "\u4F8B\uFF1A\u65B0\u5BBF\u533A\u897F\u65B0\u5BBF", en: "e.g. Shinjuku-ku, Nishi-shinjuku", vi: "VD: Shinjuku-ku", ko: "\uC608: \uC2E0\uC8FC\uCFE0\uAD6C \uB2C8\uC2DC\uC2E0\uC8FC\uCFE0", "zh-CN": "\u4F8B\uFF1A\u65B0\u5BBF\u533A\u897F\u65B0\u5BBF", "zh-TW": "\u4F8B\uFF1A\u65B0\u5BBF\u5340\u897F\u65B0\u5BBF", th: "\u0E40\u0E0A\u0E48\u0E19 \u0E0A\u0E34\u0E19\u0E08\u0E39\u0E01\u0E38", es: "ej. Shibuya-ku" },
  address2: { ja: "\u4F8B\uFF1A1-2-3", en: "e.g. 1-2-3", vi: "VD: 1-2-3", ko: "\uC608: 1-2-3", "zh-CN": "\u4F8B\uFF1A1-2-3", "zh-TW": "\u4F8B\uFF1A1-2-3", th: "\u0E40\u0E0A\u0E48\u0E19 1-2-3", es: "ej. 1-2-3" },
  address3: { ja: "\u4F8B\uFF1A\u25CB\u25CB\u30D3\u30EB 101\u53F7\u5BA4", en: "e.g. XX Building 101", vi: "VD: T\xF2a nh\xE0 XX P.101", ko: "\uC608: XX\uBE4C\uB529 101\uD638", "zh-CN": "\u4F8B\uFF1AXX\u5927\u53A6 101\u5BA4", "zh-TW": "\u4F8B\uFF1AXX\u5927\u5EC8 101\u5BA4", th: "\u0E40\u0E0A\u0E48\u0E19 \u0E15\u0E36\u0E01 XX \u0E2B\u0E49\u0E2D\u0E07 101", es: "ej. Edificio XX 101" },
  lastname: { ja: "\u4F8B\uFF1A\u7530\u4E2D", en: "e.g. Tanaka", vi: "VD: Nguy\u1EC5n", ko: "\uC608: \uB2E4\uB098\uCE74", "zh-CN": "\u4F8B\uFF1A\u7530\u4E2D", "zh-TW": "\u4F8B\uFF1A\u7530\u4E2D", th: "\u0E40\u0E0A\u0E48\u0E19 \u0E17\u0E32\u0E19\u0E32\u0E01\u0E30", es: "ej. Tanaka" },
  firstname: { ja: "\u4F8B\uFF1A\u592A\u90CE", en: "e.g. Taro", vi: "VD: V\u0103n", ko: "\uC608: \uD0C0\uB85C", "zh-CN": "\u4F8B\uFF1A\u592A\u90CE", "zh-TW": "\u4F8B\uFF1A\u592A\u90CE", th: "\u0E40\u0E0A\u0E48\u0E19 \u0E17\u0E32\u0E42\u0E23\u0E48", es: "ej. Taro" },
  kanaLastname: { ja: "\u4F8B\uFF1A\u30BF\u30CA\u30AB", en: "e.g. Tanaka", vi: "VD: Tanaka", ko: "\uC608: \uD0C0\uB098\uCE74", "zh-CN": "\u4F8B\uFF1A\u30BF\u30CA\u30AB", "zh-TW": "\u4F8B\uFF1A\u30BF\u30CA\u30AB", th: "\u0E40\u0E0A\u0E48\u0E19 \u0E17\u0E32\u0E19\u0E32\u0E01\u0E30", es: "ej. Tanaka" },
  kanaFirstname: { ja: "\u4F8B\uFF1A\u30BF\u30ED\u30A6", en: "e.g. Taro", vi: "VD: Taro", ko: "\uC608: \uD0C0\uB85C", "zh-CN": "\u4F8B\uFF1A\u30BF\u30ED\u30A6", "zh-TW": "\u4F8B\uFF1A\u30BF\u30ED\u30A6", th: "\u0E40\u0E0A\u0E48\u0E19 \u0E17\u0E32\u0E42\u0E23\u0E48", es: "ej. Taro" },
  phone: { ja: "\u4F8B\uFF1A03-1234-5678", en: "e.g. 03-1234-5678", vi: "VD: 03-1234-5678", ko: "\uC608: 03-1234-5678", "zh-CN": "\u4F8B\uFF1A03-1234-5678", "zh-TW": "\u4F8B\uFF1A03-1234-5678", th: "\u0E40\u0E0A\u0E48\u0E19 03-1234-5678", es: "ej. 03-1234-5678" },
  bankCode: { ja: "\u4F8B\uFF1A0001", en: "e.g. 0001", vi: "VD: 0001", ko: "\uC608: 0001", "zh-CN": "\u4F8B\uFF1A0001", "zh-TW": "\u4F8B\uFF1A0001", th: "\u0E40\u0E0A\u0E48\u0E19 0001", es: "ej. 0001" },
  bankName: { ja: "\u4F8B\uFF1A\u307F\u305A\u307B\u9280\u884C", en: "e.g. Mizuho Bank", vi: "VD: Ng\xE2n h\xE0ng Mizuho", ko: "\uC608: \uBBF8\uC988\uD638\uC740\uD589", "zh-CN": "\u4F8B\uFF1A\u745E\u7A57\u94F6\u884C", "zh-TW": "\u4F8B\uFF1A\u745E\u7A57\u9280\u884C", th: "\u0E40\u0E0A\u0E48\u0E19 \u0E18\u0E19\u0E32\u0E04\u0E32\u0E23\u0E21\u0E34\u0E0B\u0E39\u0E42\u0E2E", es: "ej. Banco Mizuho" },
  branchCode: { ja: "\u4F8B\uFF1A001", en: "e.g. 001", vi: "VD: 001", ko: "\uC608: 001", "zh-CN": "\u4F8B\uFF1A001", "zh-TW": "\u4F8B\uFF1A001", th: "\u0E40\u0E0A\u0E48\u0E19 001", es: "ej. 001" },
  branchName: { ja: "\u4F8B\uFF1A\u6771\u4EAC\u55B6\u696D\u90E8", en: "e.g. Tokyo Branch", vi: "VD: Chi nh\xE1nh Tokyo", ko: "\uC608: \uB3C4\uCFC4\uC601\uC5C5\uBD80", "zh-CN": "\u4F8B\uFF1A\u4E1C\u4EAC\u8425\u4E1A\u90E8", "zh-TW": "\u4F8B\uFF1A\u6771\u4EAC\u71DF\u696D\u90E8", th: "\u0E40\u0E0A\u0E48\u0E19 \u0E2A\u0E32\u0E02\u0E32\u0E42\u0E15\u0E40\u0E01\u0E35\u0E22\u0E27", es: "ej. Sucursal Tokio" },
  accountType: { ja: "\u53E3\u5EA7\u7A2E\u5225\u3092\u9078\u629E", en: "Select account type", vi: "Ch\u1ECDn lo\u1EA1i t\xE0i kho\u1EA3n", ko: "\uACC4\uC88C \uC885\uB958 \uC120\uD0DD", "zh-CN": "\u9009\u62E9\u8D26\u6237\u7C7B\u578B", "zh-TW": "\u9078\u64C7\u5E33\u6236\u985E\u578B", th: "\u0E40\u0E25\u0E37\u0E2D\u0E01\u0E1B\u0E23\u0E30\u0E40\u0E20\u0E17\u0E1A\u0E31\u0E0D\u0E0A\u0E35", es: "Seleccionar tipo" },
  accountNumber: { ja: "\u4F8B\uFF1A1234567", en: "e.g. 1234567", vi: "VD: 1234567", ko: "\uC608: 1234567", "zh-CN": "\u4F8B\uFF1A1234567", "zh-TW": "\u4F8B\uFF1A1234567", th: "\u0E40\u0E0A\u0E48\u0E19 1234567", es: "ej. 1234567" },
  accountHolder: { ja: "\u4F8B\uFF1A\u30BF\u30CA\u30AB \u30BF\u30ED\u30A6", en: "e.g. TANAKA TARO", vi: "VD: TANAKA TARO", ko: "\uC608: \uD0C0\uB098\uCE74 \uD0C0\uB85C", "zh-CN": "\u4F8B\uFF1A\u30BF\u30CA\u30AB \u30BF\u30ED\u30A6", "zh-TW": "\u4F8B\uFF1A\u30BF\u30CA\u30AB \u30BF\u30ED\u30A6", th: "\u0E40\u0E0A\u0E48\u0E19 TANAKA TARO", es: "ej. TANAKA TARO" }
};
var PREFECTURE_KEYS = [
  "hokkaido",
  "aomori",
  "iwate",
  "miyagi",
  "akita",
  "yamagata",
  "fukushima",
  "ibaraki",
  "tochigi",
  "gunma",
  "saitama",
  "chiba",
  "tokyo",
  "kanagawa",
  "niigata",
  "toyama",
  "ishikawa",
  "fukui",
  "yamanashi",
  "nagano",
  "gifu",
  "shizuoka",
  "aichi",
  "mie",
  "shiga",
  "kyoto",
  "osaka",
  "hyogo",
  "nara",
  "wakayama",
  "tottori",
  "shimane",
  "okayama",
  "hiroshima",
  "yamaguchi",
  "tokushima",
  "kagawa",
  "ehime",
  "kochi",
  "fukuoka",
  "saga",
  "nagasaki",
  "kumamoto",
  "oita",
  "miyazaki",
  "kagoshima",
  "okinawa"
];
var PREFECTURE_AREAS = {
  hokkaido: "hokkaido",
  aomori: "tohoku",
  iwate: "tohoku",
  miyagi: "tohoku",
  akita: "tohoku",
  yamagata: "tohoku",
  fukushima: "tohoku",
  ibaraki: "kanto",
  tochigi: "kanto",
  gunma: "kanto",
  saitama: "kanto",
  chiba: "kanto",
  tokyo: "kanto",
  kanagawa: "kanto",
  niigata: "chubu",
  toyama: "chubu",
  ishikawa: "chubu",
  fukui: "chubu",
  yamanashi: "chubu",
  nagano: "chubu",
  gifu: "chubu",
  shizuoka: "chubu",
  aichi: "chubu",
  mie: "kinki",
  shiga: "kinki",
  kyoto: "kinki",
  osaka: "kinki",
  hyogo: "kinki",
  nara: "kinki",
  wakayama: "kinki",
  tottori: "chugoku",
  shimane: "chugoku",
  okayama: "chugoku",
  hiroshima: "chugoku",
  yamaguchi: "chugoku",
  tokushima: "shikoku",
  kagawa: "shikoku",
  ehime: "shikoku",
  kochi: "shikoku",
  fukuoka: "kyushu",
  saga: "kyushu",
  nagasaki: "kyushu",
  kumamoto: "kyushu",
  oita: "kyushu",
  miyazaki: "kyushu",
  kagoshima: "kyushu",
  okinawa: "kyushu"
};

// src/index.ts
var PREFECTURE_LABELS = PREFECTURE_KEYS.map((key) => prefectureLabels[key]?.ja ?? key);
var PREFECTURES = PREFECTURE_LABELS;
var japanPhoneType = {
  name: "JapanesePhone",
  description: "Japanese phone number (e.g., 090-1234-5678)",
  sql: {
    sqlType: "VARCHAR",
    length: 15
  },
  typescript: {
    type: "string"
  },
  // Only base fields allowed (no type-specific fields)
  validFields: [],
  validate: (_prop) => {
    return void 0;
  }
};
var japanPostalCodeType = {
  name: "JapanesePostalCode",
  description: "Japanese postal code (e.g., 123-4567)",
  sql: {
    sqlType: "VARCHAR",
    length: 8
  },
  typescript: {
    type: "string"
  },
  validFields: []
};
var japanAddressType = {
  name: "JapaneseAddress",
  description: "Full Japanese address with postal code, prefecture (string enum), and street addresses",
  compound: true,
  validFields: [],
  expand: [
    {
      suffix: "PostalCode",
      sql: { sqlType: "VARCHAR", length: 8 },
      typescript: { type: "string" },
      rules: {
        format: "postal_code",
        minLength: 7,
        maxLength: 8
      },
      label: {
        en: "Postal Code",
        ja: "\u90F5\u4FBF\u756A\u53F7",
        vi: "M\xE3 b\u01B0u \u0111i\u1EC7n",
        ko: "\uC6B0\uD3B8\uBC88\uD638",
        "zh-CN": "\u90AE\u653F\u7F16\u7801",
        "zh-TW": "\u90F5\u905E\u5340\u865F",
        th: "\u0E23\u0E2B\u0E31\u0E2A\u0E44\u0E1B\u0E23\u0E29\u0E13\u0E35\u0E22\u0E4C",
        es: "C\xF3digo Postal"
      },
      placeholder: {
        en: "e.g. 123-4567",
        ja: "\u4F8B\uFF1A123-4567",
        vi: "VD: 123-4567",
        ko: "\uC608: 123-4567",
        "zh-CN": "\u4F8B\uFF1A123-4567",
        "zh-TW": "\u4F8B\uFF1A123-4567",
        th: "\u0E40\u0E0A\u0E48\u0E19 123-4567",
        es: "ej. 123-4567"
      }
    },
    {
      suffix: "Prefecture",
      enumRef: "Prefecture",
      // String enum: tokyo, saitama, ...
      rules: {
        minLength: 1
        // Required
      },
      label: {
        en: "Prefecture",
        ja: "\u90FD\u9053\u5E9C\u770C",
        vi: "T\u1EC9nh/Th\xE0nh ph\u1ED1",
        ko: "\uB3C4\uB3C4\uBD80\uD604",
        "zh-CN": "\u90FD\u9053\u5E9C\u53BF",
        "zh-TW": "\u90FD\u9053\u5E9C\u7E23",
        th: "\u0E08\u0E31\u0E07\u0E2B\u0E27\u0E31\u0E14",
        es: "Prefectura"
      },
      placeholder: {
        en: "Select prefecture",
        ja: "\u90FD\u9053\u5E9C\u770C\u3092\u9078\u629E",
        vi: "Ch\u1ECDn t\u1EC9nh/th\xE0nh ph\u1ED1",
        ko: "\uB3C4\uB3C4\uBD80\uD604 \uC120\uD0DD",
        "zh-CN": "\u9009\u62E9\u90FD\u9053\u5E9C\u53BF",
        "zh-TW": "\u9078\u64C7\u90FD\u9053\u5E9C\u7E23",
        th: "\u0E40\u0E25\u0E37\u0E2D\u0E01\u0E08\u0E31\u0E07\u0E2B\u0E27\u0E31\u0E14",
        es: "Seleccionar prefectura"
      }
    },
    {
      suffix: "Address1",
      sql: { sqlType: "VARCHAR", length: 255 },
      typescript: { type: "string" },
      rules: {
        minLength: 1,
        // Required
        maxLength: 255
      },
      label: {
        en: "City/Ward",
        ja: "\u5E02\u533A\u753A\u6751",
        vi: "Th\xE0nh ph\u1ED1/Qu\u1EADn",
        ko: "\uC2DC/\uAD6C/\uAD70",
        "zh-CN": "\u5E02\u533A\u753A\u6751",
        "zh-TW": "\u5E02\u5340\u753A\u6751",
        th: "\u0E40\u0E21\u0E37\u0E2D\u0E07/\u0E40\u0E02\u0E15",
        es: "Ciudad/Distrito"
      },
      placeholder: {
        en: "e.g. Shibuya-ku",
        ja: "\u4F8B\uFF1A\u6E0B\u8C37\u533A",
        vi: "VD: Qu\u1EADn Shibuya",
        ko: "\uC608: \uC2DC\uBD80\uC57C\uAD6C",
        "zh-CN": "\u4F8B\uFF1A\u6DA9\u8C37\u533A",
        "zh-TW": "\u4F8B\uFF1A\u6F80\u8C37\u5340",
        th: "\u0E40\u0E0A\u0E48\u0E19 \u0E0A\u0E34\u0E1A\u0E39\u0E22\u0E48\u0E32",
        es: "ej. Shibuya-ku"
      }
    },
    {
      suffix: "Address2",
      sql: { sqlType: "VARCHAR", length: 255 },
      typescript: { type: "string" },
      rules: {
        minLength: 1,
        // Required
        maxLength: 255
      },
      label: {
        en: "Street Address",
        ja: "\u756A\u5730",
        vi: "\u0110\u1ECBa ch\u1EC9",
        ko: "\uBC88\uC9C0",
        "zh-CN": "\u8857\u9053\u5730\u5740",
        "zh-TW": "\u8857\u9053\u5730\u5740",
        th: "\u0E17\u0E35\u0E48\u0E2D\u0E22\u0E39\u0E48",
        es: "Direcci\xF3n"
      },
      placeholder: {
        en: "e.g. 1-2-3 Dogenzaka",
        ja: "\u4F8B\uFF1A\u9053\u7384\u57421-2-3",
        vi: "VD: 1-2-3 Dogenzaka",
        ko: "\uC608: \uB3C4\uAC90\uC790\uCE74 1-2-3",
        "zh-CN": "\u4F8B\uFF1A\u9053\u7384\u57421-2-3",
        "zh-TW": "\u4F8B\uFF1A\u9053\u7384\u57421-2-3",
        th: "\u0E40\u0E0A\u0E48\u0E19 1-2-3 \u0E42\u0E14\u0E40\u0E01\u0E47\u0E19\u0E0B\u0E32\u0E01\u0E30",
        es: "ej. 1-2-3 Dogenzaka"
      }
    },
    {
      suffix: "Address3",
      sql: { sqlType: "VARCHAR", length: 255, nullable: true },
      typescript: { type: "string" },
      rules: {
        maxLength: 255
      },
      label: {
        en: "Building/Room",
        ja: "\u5EFA\u7269\u540D\u30FB\u90E8\u5C4B\u756A\u53F7",
        vi: "T\xF2a nh\xE0/Ph\xF2ng",
        ko: "\uAC74\uBB3C\uBA85/\uD638\uC2E4",
        "zh-CN": "\u5EFA\u7B51/\u623F\u95F4",
        "zh-TW": "\u5EFA\u7BC9/\u623F\u9593",
        th: "\u0E2D\u0E32\u0E04\u0E32\u0E23/\u0E2B\u0E49\u0E2D\u0E07",
        es: "Edificio/Habitaci\xF3n"
      },
      placeholder: {
        en: "e.g. ABC Building 101 (optional)",
        ja: "\u4F8B\uFF1AABC\u30D3\u30EB101\u53F7\u5BA4\uFF08\u4EFB\u610F\uFF09",
        vi: "VD: T\xF2a ABC P.101 (t\xF9y ch\u1ECDn)",
        ko: "\uC608: ABC\uBE4C\uB529 101\uD638 (\uC120\uD0DD)",
        "zh-CN": "\u4F8B\uFF1AABC\u5927\u53A6101\u5BA4\uFF08\u53EF\u9009\uFF09",
        "zh-TW": "\u4F8B\uFF1AABC\u5927\u5EC8101\u5BA4\uFF08\u53EF\u9078\uFF09",
        th: "\u0E40\u0E0A\u0E48\u0E19 \u0E2D\u0E32\u0E04\u0E32\u0E23 ABC 101 (\u0E44\u0E21\u0E48\u0E1A\u0E31\u0E07\u0E04\u0E31\u0E1A)",
        es: "ej. Edificio ABC 101 (opcional)"
      }
    }
  ],
  accessors: [
    {
      name: "full_address",
      fields: ["Address1", "Address2", "Address3"],
      separator: ""
    }
  ]
};
var japanNameType = {
  name: "JapaneseName",
  description: "Japanese name with kanji and kana variants",
  compound: true,
  validFields: [],
  expand: [
    {
      suffix: "Lastname",
      sql: { sqlType: "VARCHAR", length: 50 },
      typescript: { type: "string" },
      rules: {
        minLength: 1,
        // Required
        maxLength: 50
      },
      label: {
        en: "Last Name",
        ja: "\u59D3",
        vi: "H\u1ECD",
        ko: "\uC131",
        "zh-CN": "\u59D3",
        "zh-TW": "\u59D3",
        th: "\u0E19\u0E32\u0E21\u0E2A\u0E01\u0E38\u0E25",
        es: "Apellido"
      },
      placeholder: {
        en: "e.g. Yamada",
        ja: "\u4F8B\uFF1A\u5C71\u7530",
        vi: "VD: Nguy\u1EC5n",
        ko: "\uC608: \uC57C\uB9C8\uB2E4",
        "zh-CN": "\u4F8B\uFF1A\u5C71\u7530",
        "zh-TW": "\u4F8B\uFF1A\u5C71\u7530",
        th: "\u0E40\u0E0A\u0E48\u0E19 \u0E22\u0E32\u0E21\u0E32\u0E14\u0E30",
        es: "ej. Yamada"
      }
    },
    {
      suffix: "Firstname",
      sql: { sqlType: "VARCHAR", length: 50 },
      typescript: { type: "string" },
      rules: {
        minLength: 1,
        // Required
        maxLength: 50
      },
      label: {
        en: "First Name",
        ja: "\u540D",
        vi: "T\xEAn",
        ko: "\uC774\uB984",
        "zh-CN": "\u540D",
        "zh-TW": "\u540D",
        th: "\u0E0A\u0E37\u0E48\u0E2D",
        es: "Nombre"
      },
      placeholder: {
        en: "e.g. Taro",
        ja: "\u4F8B\uFF1A\u592A\u90CE",
        vi: "VD: V\u0103n A",
        ko: "\uC608: \uD0C0\uB85C",
        "zh-CN": "\u4F8B\uFF1A\u592A\u90CE",
        "zh-TW": "\u4F8B\uFF1A\u592A\u90CE",
        th: "\u0E40\u0E0A\u0E48\u0E19 \u0E17\u0E32\u0E42\u0E23\u0E48",
        es: "ej. Taro"
      }
    },
    {
      suffix: "KanaLastname",
      sql: { sqlType: "VARCHAR", length: 100, nullable: true },
      typescript: { type: "string" },
      rules: {
        maxLength: 100,
        pattern: "^[\u30A1-\u30F6\u30FC\u30FB\u3000]+$"
        // Katakana only
      },
      label: {
        en: "Last Name (Kana)",
        ja: "\u59D3\uFF08\u30AB\u30CA\uFF09",
        vi: "H\u1ECD (Kana)",
        ko: "\uC131 (\uAC00\uB098)",
        "zh-CN": "\u59D3\uFF08\u7247\u5047\u540D\uFF09",
        "zh-TW": "\u59D3\uFF08\u7247\u5047\u540D\uFF09",
        th: "\u0E19\u0E32\u0E21\u0E2A\u0E01\u0E38\u0E25 (\u0E04\u0E32\u0E19\u0E30)",
        es: "Apellido (Kana)"
      },
      placeholder: {
        en: "e.g. Yamada (Katakana)",
        ja: "\u4F8B\uFF1A\u30E4\u30DE\u30C0",
        vi: "VD: Yamada (Katakana)",
        ko: "\uC608: \uC57C\uB9C8\uB2E4 (\uAC00\uD0C0\uCE74\uB098)",
        "zh-CN": "\u4F8B\uFF1A\u30E4\u30DE\u30C0",
        "zh-TW": "\u4F8B\uFF1A\u30E4\u30DE\u30C0",
        th: "\u0E40\u0E0A\u0E48\u0E19 \u30E4\u30DE\u30C0",
        es: "ej. \u30E4\u30DE\u30C0"
      }
    },
    {
      suffix: "KanaFirstname",
      sql: { sqlType: "VARCHAR", length: 100, nullable: true },
      typescript: { type: "string" },
      rules: {
        maxLength: 100,
        pattern: "^[\u30A1-\u30F6\u30FC\u30FB\u3000]+$"
        // Katakana only
      },
      label: {
        en: "First Name (Kana)",
        ja: "\u540D\uFF08\u30AB\u30CA\uFF09",
        vi: "T\xEAn (Kana)",
        ko: "\uC774\uB984 (\uAC00\uB098)",
        "zh-CN": "\u540D\uFF08\u7247\u5047\u540D\uFF09",
        "zh-TW": "\u540D\uFF08\u7247\u5047\u540D\uFF09",
        th: "\u0E0A\u0E37\u0E48\u0E2D (\u0E04\u0E32\u0E19\u0E30)",
        es: "Nombre (Kana)"
      },
      placeholder: {
        en: "e.g. Taro (Katakana)",
        ja: "\u4F8B\uFF1A\u30BF\u30ED\u30A6",
        vi: "VD: Van A (Katakana)",
        ko: "\uC608: \uD0C0\uB85C (\uAC00\uD0C0\uCE74\uB098)",
        "zh-CN": "\u4F8B\uFF1A\u30BF\u30ED\u30A6",
        "zh-TW": "\u4F8B\uFF1A\u30BF\u30ED\u30A6",
        th: "\u0E40\u0E0A\u0E48\u0E19 \u30BF\u30ED\u30A6",
        es: "ej. \u30BF\u30ED\u30A6"
      }
    }
  ],
  accessors: [
    {
      name: "full_name",
      fields: ["Lastname", "Firstname"],
      separator: " "
    },
    {
      name: "full_name_kana",
      fields: ["KanaLastname", "KanaFirstname"],
      separator: " "
    }
  ]
};
var japanBankAccountType = {
  name: "JapaneseBankAccount",
  description: "Japanese bank account with bank/branch codes and names, type, number, and holder name",
  compound: true,
  validFields: [],
  expand: [
    {
      suffix: "BankCode",
      sql: { sqlType: "VARCHAR", length: 4, nullable: true },
      typescript: { type: "string" },
      rules: {
        minLength: 4,
        maxLength: 4,
        pattern: "^[0-9]{4}$"
        // 4 digits
      },
      label: {
        en: "Bank Code",
        ja: "\u9280\u884C\u30B3\u30FC\u30C9",
        vi: "M\xE3 ng\xE2n h\xE0ng",
        ko: "\uC740\uD589\uCF54\uB4DC",
        "zh-CN": "\u94F6\u884C\u4EE3\u7801",
        "zh-TW": "\u9280\u884C\u4EE3\u78BC",
        th: "\u0E23\u0E2B\u0E31\u0E2A\u0E18\u0E19\u0E32\u0E04\u0E32\u0E23",
        es: "C\xF3digo Bancario"
      },
      placeholder: {
        en: "e.g. 0001",
        ja: "\u4F8B\uFF1A0001",
        vi: "VD: 0001",
        ko: "\uC608: 0001",
        "zh-CN": "\u4F8B\uFF1A0001",
        "zh-TW": "\u4F8B\uFF1A0001",
        th: "\u0E40\u0E0A\u0E48\u0E19 0001",
        es: "ej. 0001"
      }
    },
    {
      suffix: "BankName",
      sql: { sqlType: "VARCHAR", length: 100, nullable: true },
      typescript: { type: "string" },
      rules: {
        maxLength: 100
      },
      label: {
        en: "Bank Name",
        ja: "\u9280\u884C\u540D",
        vi: "T\xEAn ng\xE2n h\xE0ng",
        ko: "\uC740\uD589\uBA85",
        "zh-CN": "\u94F6\u884C\u540D\u79F0",
        "zh-TW": "\u9280\u884C\u540D\u7A31",
        th: "\u0E0A\u0E37\u0E48\u0E2D\u0E18\u0E19\u0E32\u0E04\u0E32\u0E23",
        es: "Nombre del Banco"
      },
      placeholder: {
        en: "e.g. MUFG Bank",
        ja: "\u4F8B\uFF1A\u4E09\u83F1UFJ\u9280\u884C",
        vi: "VD: Ng\xE2n h\xE0ng MUFG",
        ko: "\uC608: \uBBF8\uC4F0\uBE44\uC2DCUFJ\uC740\uD589",
        "zh-CN": "\u4F8B\uFF1A\u4E09\u83F1UFJ\u94F6\u884C",
        "zh-TW": "\u4F8B\uFF1A\u4E09\u83F1UFJ\u9280\u884C",
        th: "\u0E40\u0E0A\u0E48\u0E19 \u0E18\u0E19\u0E32\u0E04\u0E32\u0E23 MUFG",
        es: "ej. Banco MUFG"
      }
    },
    {
      suffix: "BranchCode",
      sql: { sqlType: "VARCHAR", length: 3, nullable: true },
      typescript: { type: "string" },
      rules: {
        minLength: 3,
        maxLength: 3,
        pattern: "^[0-9]{3}$"
        // 3 digits
      },
      label: {
        en: "Branch Code",
        ja: "\u652F\u5E97\u30B3\u30FC\u30C9",
        vi: "M\xE3 chi nh\xE1nh",
        ko: "\uC9C0\uC810\uCF54\uB4DC",
        "zh-CN": "\u652F\u884C\u4EE3\u7801",
        "zh-TW": "\u5206\u884C\u4EE3\u78BC",
        th: "\u0E23\u0E2B\u0E31\u0E2A\u0E2A\u0E32\u0E02\u0E32",
        es: "C\xF3digo de Sucursal"
      },
      placeholder: {
        en: "e.g. 001",
        ja: "\u4F8B\uFF1A001",
        vi: "VD: 001",
        ko: "\uC608: 001",
        "zh-CN": "\u4F8B\uFF1A001",
        "zh-TW": "\u4F8B\uFF1A001",
        th: "\u0E40\u0E0A\u0E48\u0E19 001",
        es: "ej. 001"
      }
    },
    {
      suffix: "BranchName",
      sql: { sqlType: "VARCHAR", length: 100, nullable: true },
      typescript: { type: "string" },
      rules: {
        maxLength: 100
      },
      label: {
        en: "Branch Name",
        ja: "\u652F\u5E97\u540D",
        vi: "T\xEAn chi nh\xE1nh",
        ko: "\uC9C0\uC810\uBA85",
        "zh-CN": "\u652F\u884C\u540D\u79F0",
        "zh-TW": "\u5206\u884C\u540D\u7A31",
        th: "\u0E0A\u0E37\u0E48\u0E2D\u0E2A\u0E32\u0E02\u0E32",
        es: "Nombre de Sucursal"
      },
      placeholder: {
        en: "e.g. Shibuya Branch",
        ja: "\u4F8B\uFF1A\u6E0B\u8C37\u652F\u5E97",
        vi: "VD: Chi nh\xE1nh Shibuya",
        ko: "\uC608: \uC2DC\uBD80\uC57C\uC9C0\uC810",
        "zh-CN": "\u4F8B\uFF1A\u6DA9\u8C37\u652F\u884C",
        "zh-TW": "\u4F8B\uFF1A\u6F80\u8C37\u5206\u884C",
        th: "\u0E40\u0E0A\u0E48\u0E19 \u0E2A\u0E32\u0E02\u0E32\u0E0A\u0E34\u0E1A\u0E39\u0E22\u0E48\u0E32",
        es: "ej. Sucursal Shibuya"
      }
    },
    {
      suffix: "AccountType",
      sql: { sqlType: "VARCHAR", length: 1, nullable: true },
      enumRef: "BankAccountType",
      // 普通/当座/貯蓄
      label: {
        en: "Account Type",
        ja: "\u53E3\u5EA7\u7A2E\u5225",
        vi: "Lo\u1EA1i t\xE0i kho\u1EA3n",
        ko: "\uACC4\uC88C\uC885\uB958",
        "zh-CN": "\u8D26\u6237\u7C7B\u578B",
        "zh-TW": "\u5E33\u6236\u985E\u578B",
        th: "\u0E1B\u0E23\u0E30\u0E40\u0E20\u0E17\u0E1A\u0E31\u0E0D\u0E0A\u0E35",
        es: "Tipo de Cuenta"
      },
      placeholder: {
        en: "Select account type",
        ja: "\u53E3\u5EA7\u7A2E\u5225\u3092\u9078\u629E",
        vi: "Ch\u1ECDn lo\u1EA1i t\xE0i kho\u1EA3n",
        ko: "\uACC4\uC88C\uC885\uB958 \uC120\uD0DD",
        "zh-CN": "\u9009\u62E9\u8D26\u6237\u7C7B\u578B",
        "zh-TW": "\u9078\u64C7\u5E33\u6236\u985E\u578B",
        th: "\u0E40\u0E25\u0E37\u0E2D\u0E01\u0E1B\u0E23\u0E30\u0E40\u0E20\u0E17\u0E1A\u0E31\u0E0D\u0E0A\u0E35",
        es: "Seleccionar tipo de cuenta"
      }
    },
    {
      suffix: "AccountNumber",
      sql: { sqlType: "VARCHAR", length: 7, nullable: true },
      typescript: { type: "string" },
      rules: {
        minLength: 7,
        maxLength: 7,
        pattern: "^[0-9]{7}$"
        // 7 digits
      },
      label: {
        en: "Account Number",
        ja: "\u53E3\u5EA7\u756A\u53F7",
        vi: "S\u1ED1 t\xE0i kho\u1EA3n",
        ko: "\uACC4\uC88C\uBC88\uD638",
        "zh-CN": "\u8D26\u6237\u53F7\u7801",
        "zh-TW": "\u5E33\u6236\u865F\u78BC",
        th: "\u0E40\u0E25\u0E02\u0E1A\u0E31\u0E0D\u0E0A\u0E35",
        es: "N\xFAmero de Cuenta"
      },
      placeholder: {
        en: "e.g. 1234567",
        ja: "\u4F8B\uFF1A1234567",
        vi: "VD: 1234567",
        ko: "\uC608: 1234567",
        "zh-CN": "\u4F8B\uFF1A1234567",
        "zh-TW": "\u4F8B\uFF1A1234567",
        th: "\u0E40\u0E0A\u0E48\u0E19 1234567",
        es: "ej. 1234567"
      }
    },
    {
      suffix: "AccountHolder",
      sql: { sqlType: "VARCHAR", length: 100, nullable: true },
      typescript: { type: "string" },
      rules: {
        maxLength: 100,
        pattern: "^[\u30A1-\u30F6\u30FC\u30FB\u3000\\s]+$"
        // Katakana + space
      },
      label: {
        en: "Account Holder",
        ja: "\u53E3\u5EA7\u540D\u7FA9",
        vi: "T\xEAn ch\u1EE7 t\xE0i kho\u1EA3n",
        ko: "\uC608\uAE08\uC8FC",
        "zh-CN": "\u8D26\u6237\u6301\u6709\u4EBA",
        "zh-TW": "\u5E33\u6236\u6301\u6709\u4EBA",
        th: "\u0E0A\u0E37\u0E48\u0E2D\u0E40\u0E08\u0E49\u0E32\u0E02\u0E2D\u0E07\u0E1A\u0E31\u0E0D\u0E0A\u0E35",
        es: "Titular de la Cuenta"
      },
      placeholder: {
        en: "e.g. Yamada Taro (Katakana)",
        ja: "\u4F8B\uFF1A\u30E4\u30DE\u30C0 \u30BF\u30ED\u30A6",
        vi: "VD: Yamada Taro (Katakana)",
        ko: "\uC608: \uC57C\uB9C8\uB2E4 \uD0C0\uB85C (\uAC00\uD0C0\uCE74\uB098)",
        "zh-CN": "\u4F8B\uFF1A\u30E4\u30DE\u30C0 \u30BF\u30ED\u30A6",
        "zh-TW": "\u4F8B\uFF1A\u30E4\u30DE\u30C0 \u30BF\u30ED\u30A6",
        th: "\u0E40\u0E0A\u0E48\u0E19 \u30E4\u30DE\u30C0 \u30BF\u30ED\u30A6",
        es: "ej. \u30E4\u30DE\u30C0 \u30BF\u30ED\u30A6"
      }
    }
  ]
};
var bankAccountTypeEnum = {
  name: "BankAccountType",
  displayName: enumDisplayNames.BankAccountType,
  description: "Japanese bank account types (Zengin system codes)",
  values: [
    { value: "1", label: bankAccountTypeLabels["1"] },
    { value: "2", label: bankAccountTypeLabels["2"] },
    { value: "4", label: bankAccountTypeLabels["4"] }
  ]
};
var prefectureEnum = {
  name: "Prefecture",
  displayName: enumDisplayNames.Prefecture,
  description: "Japanese prefectures (string enum)",
  values: PREFECTURE_KEYS.map((key, index) => ({
    value: key,
    label: prefectureLabels[key],
    extra: { area: PREFECTURE_AREAS[key], code: index + 1 }
  }))
};
var prefectureCodeEnum = {
  name: "PrefectureCode",
  displayName: enumDisplayNames.PrefectureCode,
  description: "Japanese prefectures (JIS X 0401 numeric codes)",
  values: PREFECTURE_KEYS.map((key, index) => ({
    value: String(index + 1),
    label: prefectureLabels[key],
    extra: { area: PREFECTURE_AREAS[key] }
  }))
};
var japanTypesPlugin = {
  name: "@famgia/omnify-japan",
  version: "0.0.1",
  types: [
    japanPhoneType,
    japanPostalCodeType,
    japanAddressType,
    japanNameType,
    japanBankAccountType
  ],
  enums: [bankAccountTypeEnum, prefectureEnum, prefectureCodeEnum],
  setup: (context) => {
    context.logger.debug("Japan types plugin loaded");
  }
};
var index_default = japanTypesPlugin;
export {
  PREFECTURES,
  PREFECTURE_AREAS,
  PREFECTURE_KEYS,
  PREFECTURE_LABELS,
  bankAccountTypeLabels,
  index_default as default,
  enumDisplayNames,
  fieldLabels,
  japanTypesPlugin,
  placeholders,
  prefectureAreaLabels,
  prefectureLabels
};
