// @famgia/omnify-types@2.0.3/dist/index.js
// src/errors.ts
function ok(value) {
  return { ok: true, value };
}
function err(error) {
  return { ok: false, error };
}

// src/i18n.ts
var DEFAULT_LOCALE_CONFIG = {
  locales: ["en"],
  defaultLocale: "en"
};
function isLocaleMap(value) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return false;
  }
  return Object.values(value).every((v) => typeof v === "string");
}
function isLocalizedString(value) {
  return typeof value === "string" || isLocaleMap(value);
}
function resolveLocalizedString(value, options = {}) {
  if (value === void 0 || value === null) {
    return void 0;
  }
  if (typeof value === "string") {
    return value;
  }
  const config = options.config ?? DEFAULT_LOCALE_CONFIG;
  const requestedLocale = options.locale ?? config.defaultLocale;
  if (requestedLocale in value) {
    return value[requestedLocale];
  }
  if (config.fallbackLocale && config.fallbackLocale in value) {
    return value[config.fallbackLocale];
  }
  if (config.defaultLocale in value) {
    return value[config.defaultLocale];
  }
  const availableLocales = Object.keys(value);
  const firstLocale = availableLocales[0];
  if (firstLocale !== void 0) {
    return value[firstLocale];
  }
  if (options.strict) {
    throw new Error(
      `Locale '${requestedLocale}' not found in localized string. Available locales: none`
    );
  }
  return void 0;
}
function getAvailableLocales(value) {
  if (value === void 0 || value === null) {
    return [];
  }
  if (typeof value === "string") {
    return ["default"];
  }
  return Object.keys(value);
}
function toLocaleMap(value, defaultLocale = "en") {
  if (value === void 0 || value === null) {
    return void 0;
  }
  if (typeof value === "string") {
    return { [defaultLocale]: value };
  }
  return value;
}
function mergeLocalizedStrings(base, override, defaultLocale = "en") {
  const baseMap = toLocaleMap(base, defaultLocale);
  const overrideMap = toLocaleMap(override, defaultLocale);
  if (!baseMap && !overrideMap) {
    return void 0;
  }
  return {
    ...baseMap,
    ...overrideMap
  };
}
export {
  DEFAULT_LOCALE_CONFIG,
  err,
  getAvailableLocales,
  isLocaleMap,
  isLocalizedString,
  mergeLocalizedStrings,
  ok,
  resolveLocalizedString,
  toLocaleMap
};
