// @famgia/omnify-types@2.0.3/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  DEFAULT_LOCALE_CONFIG: () => DEFAULT_LOCALE_CONFIG,
  err: () => err,
  getAvailableLocales: () => getAvailableLocales,
  isLocaleMap: () => isLocaleMap,
  isLocalizedString: () => isLocalizedString,
  mergeLocalizedStrings: () => mergeLocalizedStrings,
  ok: () => ok,
  resolveLocalizedString: () => resolveLocalizedString,
  toLocaleMap: () => toLocaleMap
});
module.exports = __toCommonJS(index_exports);

// src/errors.ts
function ok(value) {
  return { ok: true, value };
}
function err(error) {
  return { ok: false, error };
}

// src/i18n.ts
var DEFAULT_LOCALE_CONFIG = {
  locales: ["en"],
  defaultLocale: "en"
};
function isLocaleMap(value) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return false;
  }
  return Object.values(value).every((v) => typeof v === "string");
}
function isLocalizedString(value) {
  return typeof value === "string" || isLocaleMap(value);
}
function resolveLocalizedString(value, options = {}) {
  if (value === void 0 || value === null) {
    return void 0;
  }
  if (typeof value === "string") {
    return value;
  }
  const config = options.config ?? DEFAULT_LOCALE_CONFIG;
  const requestedLocale = options.locale ?? config.defaultLocale;
  if (requestedLocale in value) {
    return value[requestedLocale];
  }
  if (config.fallbackLocale && config.fallbackLocale in value) {
    return value[config.fallbackLocale];
  }
  if (config.defaultLocale in value) {
    return value[config.defaultLocale];
  }
  const availableLocales = Object.keys(value);
  const firstLocale = availableLocales[0];
  if (firstLocale !== void 0) {
    return value[firstLocale];
  }
  if (options.strict) {
    throw new Error(
      `Locale '${requestedLocale}' not found in localized string. Available locales: none`
    );
  }
  return void 0;
}
function getAvailableLocales(value) {
  if (value === void 0 || value === null) {
    return [];
  }
  if (typeof value === "string") {
    return ["default"];
  }
  return Object.keys(value);
}
function toLocaleMap(value, defaultLocale = "en") {
  if (value === void 0 || value === null) {
    return void 0;
  }
  if (typeof value === "string") {
    return { [defaultLocale]: value };
  }
  return value;
}
function mergeLocalizedStrings(base, override, defaultLocale = "en") {
  const baseMap = toLocaleMap(base, defaultLocale);
  const overrideMap = toLocaleMap(override, defaultLocale);
  if (!baseMap && !overrideMap) {
    return void 0;
  }
  return {
    ...baseMap,
    ...overrideMap
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_LOCALE_CONFIG,
  err,
  getAvailableLocales,
  isLocaleMap,
  isLocalizedString,
  mergeLocalizedStrings,
  ok,
  resolveLocalizedString,
  toLocaleMap
});
