// @omnify-base/schemas/Branch.ts
/**
 * ⚠️ DO NOT EDIT THIS FILE! ⚠️
 * このファイルを編集しないでください！
 * KHÔNG ĐƯỢC SỬA FILE NÀY!
 *
 * Auto-generated TypeScript types from Omnify schemas.
 * Any manual changes will be OVERWRITTEN on next generation.
 *
 * To modify: Edit the schema YAML file and run: npx omnify generate
 */

import { z } from 'zod';
import type { DateTimeString } from './common';

/**
 * 支店
 */
export interface Branch {
  /** Primary key */
  id: string;
  /** Console Branch ID */
  console_branch_id: unknown;
  /** Console Organization ID */
  console_org_id: unknown;
  /** 支店コード */
  code: string;
  /** 支店名 */
  name: string;
  /** Creation timestamp */
  created_at?: DateTimeString;
  /** Last update timestamp */
  updated_at?: DateTimeString;
  /** Soft delete timestamp */
  deleted_at?: DateTimeString;
}

// ============================================================================
// I18n (Internationalization)
// ============================================================================

/**
 * Unified i18n object for Branch
 * Contains model label and all field labels/placeholders
 */
export const branchI18n = {
  /** Model display name */
  label: {"ja":"支店","en":"Branch","vi":"Chi nhánh"},
  /** Field labels and placeholders */
  fields: {
    console_branch_id: {
      label: {"ja":"Console Branch ID","en":"Console Branch ID","vi":"Console Branch ID"},
    },
    console_org_id: {
      label: {"ja":"Console Organization ID","en":"Console Organization ID","vi":"Console Organization ID"},
    },
    code: {
      label: {"ja":"支店コード","en":"Branch Code","vi":"Mã chi nhánh"},
    },
    name: {
      label: {"ja":"支店名","en":"Branch Name","vi":"Tên chi nhánh"},
    },
  },
} as const;

// ============================================================================
// Zod Schemas
// ============================================================================

/** Field schemas for Branch */
export const baseBranchSchemas = {
  console_branch_id: z.string(),
  console_org_id: z.string(),
  code: z.string().min(1).max(20),
  name: z.string().min(1).max(100),
} as const;

/** Create schema for Branch (POST requests) */
export const baseBranchCreateSchema = z.object({
  console_branch_id: baseBranchSchemas.console_branch_id,
  console_org_id: baseBranchSchemas.console_org_id,
  code: baseBranchSchemas.code,
  name: baseBranchSchemas.name,
});

/** Update schema for Branch (PUT/PATCH requests) */
export const baseBranchUpdateSchema = baseBranchCreateSchema.partial();

// ============================================================================
// Inferred Types
// ============================================================================

export type BaseBranchCreate = z.infer<typeof baseBranchCreateSchema>;
export type BaseBranchUpdate = z.infer<typeof baseBranchUpdateSchema>;

// ============================================================================
// I18n Helper Functions
// ============================================================================

/** Get model label for a specific locale */
export function getBranchLabel(locale: string): string {
  return branchI18n.label[locale as keyof typeof branchI18n.label] ?? branchI18n.label['en'] ?? 'Branch';
}

/** Get field label for a specific locale */
export function getBranchFieldLabel(field: string, locale: string): string {
  const fieldI18n = branchI18n.fields[field as keyof typeof branchI18n.fields];
  if (!fieldI18n) return field;
  return fieldI18n.label[locale as keyof typeof fieldI18n.label] ?? fieldI18n.label['en'] ?? field;
}

/** Get field placeholder for a specific locale */
export function getBranchFieldPlaceholder(field: string, locale: string): string {
  const fieldI18n = branchI18n.fields[field as keyof typeof branchI18n.fields];
  if (!fieldI18n || !('placeholder' in fieldI18n)) return '';
  const placeholder = fieldI18n.placeholder as Record<string, string>;
  return placeholder[locale] ?? placeholder['en'] ?? '';
}
